packages/verified-fetch/README.md

```diff
@@ -658,6 +658,198 @@ Known Errors that can be thrown:
+    (for example, by performing path walking or decoding) and the pipeline should continue.
+  - **Throw a `PluginError`**: This logs a non-fatal error and continues the pipeline.
+  - **Throw a `PluginFatalError`**: This logs a fatal error and stops the pipeline immediately.
+
+Plugins are executed in a chain (a **plugin pipeline**):
+
+1. **Initialization:**
+   - The `VerifiedFetch` class is instantiated with a list of plugins.
+   - When a request is made via the `fetch` method, the resource and options are parsed to
+     create a mutable `PluginContext` object.
+
+2. **Pipeline Execution:**
+
+   - The pipeline repeatedly checks, up to a maximum number of passes (default = 3), which plugins
+     are currently able to handle the request by calling each plugin’s `canHandle()` method.
+   - Plugins that have not yet been called in the current run and return `true` for `canHandle()`
+     are invoked in sequence.
+     - If a plugin returns a final `Response` or throws a `PluginFatalError`, the pipeline immediately
+       stops and that response is returned.
+     - If a plugin returns `undefined`, it is assumed to have updated the context (for example, by
+       performing path walking), and the pipeline proceeds to a new pass where plugins are re‑evaluated.
+   - If no plugin modifies the context (i.e. no change to `context.modified`) and no final response is
+     produced after iterating through all plugins, the pipeline exits and a default “Not Supported”
+     response is returned.
+
+   **Diagram of the Plugin Pipeline:**
+
+```mermaid
+flowchart TD
+    A[Resource & Options] --> B[Parse into PluginContext]
+    B --> C[Plugin Pipeline]
+    subgraph IP[Iterative Passes max 3 passes]
+      C1[Check canHandle for each plugin]
+      C2[Call handle on ready plugins]
+      C3[Update PluginContext if partial work is done]
+      C1 --> C2
+      C2 --> C3
+    end
+    C --> IP
+    IP --> D[Final Response]
+```
+
+3. **Finalization:**
+   - After the pipeline completes, the resulting response & context is processed (e.g. headers such as ETag,
+     Cache‑Control, and Content‑Disposition are set) and returned.
+
+Please see the original discussion on extensibility in [Issue #167](https://github.com/ipfs/helia-verified-fetch/issues/167).
+
+***
+
+### Extending Verified‑Fetch with Custom Plugins
+
+To add your own plugin:
+
+1. **Extend the BasePlugin:**
+
+   Create a new class that extends `BasePlugin` and implements:
+
+   - `canHandle(context: PluginContext): boolean`
+   - `handle(context: PluginContext): Promise<Response | null>`
+
+## Example - custom plugin
+
+```typescript
+   import { BasePlugin, type PluginContext, type VerifiedFetchPluginFactory, type PluginOptions } from '@helia/verified-fetch'
+   import { okResponse } from './dist/src/utils/responses.js'
+
+   export class MyCustomPlugin extends BasePlugin {
+     // Optionally, list any codec codes your plugin supports:
```

Comment:
What is the code actually used for?

Reply:
i'm not sure I understand.

Reply:
oh.. codec codes. Got it.

Currently, they are used to express an expectation of particular codecs a plugin can handle.. (for the fallback discussed in https://github.com/ipfs/helia-verified-fetch/pull/169#discussion_r1952692616). In the future this would be unnecessary and fully encapsulated within the `canHandle` functions, or transformer/terminal plugins